File: src/1-500/087/ScrambleString.java
class Solution {
    public boolean isScramble(String s1, String s2) {
		if (s1 == null || s2 == null || s1.length() != s2.length()) {
			return false;
		}
		if (s1.equals(s2)) {
			return true;
		}
		
		int len = s1.length();
		boolean[][][] map = new boolean[len+1][len+1][len+1];
		
		for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                map[i][j][1] = (s1.charAt(i) == s2.charAt(j));
            }
		}
		for (int n=2; n<=len; n++) {
			for (int i=0; i<=len-n; i++) {
				for (int j=0; j<=len-n; j++) {
					for (int k=1; k<=n && !map[i][j][n]; k++) {
						map[i][j][n] = (map[i][j][k] && map[i+k][j+k][n-k]) 
							|| (map[i][j+n-k][k] && map[i+k][j][n-k]);
					}
				}
			}
		}
		
		return map[0][0][len];
	}
}

File: src/1-500/087/ScrambleString.py
class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        if s1 is None or s2 is None or len(s1) != len(s2):
            return False

        Len= len(s1)
        if Len == 1:
            return s1 == s2
        else: 
            for i in range(1, Len):
                s1_left = s1[:i]
                s1_right = s1[i:]
                s2_left = s2[:-i]
                s2_right = s2[-i:]
            
                if self.isScramble(s1_left, s2_right) and self.isScramble(s1_right, s2_left):
                    return True
                
                s2_left = s2[:i]
                s2_right = s2[i:]
                if self.isScramble(s1_left, s2_left) and self.isScramble(s1_right, s2_right):
                    return True
            return False